include/tvm/target/target_device.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/target/target_device.h
+ * \brief A compile time representation of a target device.
+ *
+ * This data structure consists of both the compiler target and a virtual device,
+ * a tvm::Device where the the identifier is a virtual identifier and a concrete
+ * device type.
+ *
+ * Executors are required to handle how to map virtual device identifiers to physical
+ * device identifiers.
+ *
+ * The reason to introduce this data structure is that for much of compilation we
```

Comment:
I kind of agree with this, but it requires a lot of other contextual information to understand things like the reach and purpose of this data structure. Specifically, it would be great to assume the reader does not understand the distinction between "concrete device" and "target" here (or explain it).

I think you could say something like:
```
This data structure defines a compile-time identifier for one of the devices specified in the Target. For instance, a Target string `cuda --target-host=llvm -mcpu=i386` requires the compiler to track two such devices: the GPU used by CUDA and the CPU targeted by LLVM. Before this data structure, the two different devices were configured by the runtime by finding a user-specified `DLDevice` whose `device_type` field matches the Target. More complex deployment scenarios can't be modeled coherently across compiler and runtime in this system without aliasing in the target string:
 * two of any `device_type` with differing capabilities
 * reconfigurable devices (e.g. FPGA) for which TVM requires two distinct configurations
 * devices with memory constraints (which need to be expressed in terms of a specific, but unresolved-at-compile-time DLDevice)
```

Reply:
BTW I owe you all an RFC to both justify this and bring together the threads on target maps, default targets, and unified target/device/memory scope planning.